arch/x86/kvm/x86.c

```diff
@@ -10032,6 +10037,17 @@ static int vcpu_run(struct kvm_vcpu *vcpu)
 				return r;
 			vcpu->srcu_idx = srcu_read_lock(&kvm->srcu);
 		}
+
+		/*
+		 * exit to userspace when dirty quota is full (if dirty quota
+		 * throttling is enabled)
+		 */
+		if (atomic_read(vcpu->kvm->dirty_quota_throttling) &&
+				is_dirty_quota_full(vcpu)) {
+			vcpu->run->exit_reason = KVM_EXIT_DIRTY_QUOTA_FULL;
+			r = 0;
+			break;
+		}
```

Comment:
This obviously needs a few more changes as we need to pass the dirty counter to the userspace (to do the adjustments in case of over dirtying as we saw in some cases) and also we need to update the quota with whatever returned from userspace. Sean (the reviewer) suggested to use the kvm_run structure to do this but this surely needs more discussion.